src/sherpa_ai/utils.py

```diff
@@ -325,3 +330,217 @@ def check_if_number_exist(result: str, source: str):
+          """
+    )
+
+    llm_result = llm.predict(instruction)
+
+    checkup_json = json_extractor(llm_result)
+    return checkup_json
+
+
+def text_similarity_by_metrics(check_entity: List[str], source_entity: List[str]):
+    """
+        Check entity similarity based on Jaccard and Levenshtein metrics.
+
+        Args:
+        - check_entity (List[str]): List of entities to check.
+        - source_entity (List[str]): List of reference entities.
+
+        Returns:
+        dict: Result of the check containing 'entity_exist' and 'messages'.
+    """
+
+    threshold = 0.75
+    error_entity = []
+    message = ""
+    levenshtein_constant = 0.5
+    for entity in source_entity:
+        metrics_value = 0
+        for sub_entitiy in check_entity:
+            word1 = entity
+            word2 = sub_entitiy
+
+            combined_similarity = string_comparison_with_jaccard_and_levenshtein(
+                word1=word1, word2=word2, levenshtein_constant=levenshtein_constant
+            )
+            if metrics_value < combined_similarity:
+                metrics_value = combined_similarity
+        if metrics_value < threshold:
+            error_entity.append(entity)
+
+    if len(error_entity) > 0:
+        for entity in error_entity:
+            message += entity + ", "
+        message = f"remember to address these entities {message} in final the answer."
+        return {"entity_exist": False, "messages": message}
+    return {"entity_exist": True, "messages": message}
+
+
+def text_similarity(check_entity: List[str], source_entity: List[str]):
+    """
+        Check if entities from a reference list are present in another list.
+
+        Args:
+        - check_entity ([str]): List of entities to check.
+        - source_entity ([str]): List of reference entities.
+
+        Returns:
+        dict: Result of the check containing 'entity_exist' and 'messages'.
+    """
+
+    error_entity = []
+    message = ""
+    for entity in source_entity:
+        print(entity)
+        if entity not in check_entity:
+            error_entity.append(entity)
+    if len(error_entity) > 0:
+        for entity in error_entity:
+            message += entity + ", "
+        message = f"remember to address these entities {message} in final the answer."
+        return {"entity_exist": False, "messages": message}
+    return {"entity_exist": True, "messages": message}
+
+
+def check_entities_match(result: str, source: str, stage: int):
+    """
+        Check if entities extracted from a question are present in an answer.
+
+        Args:
+        - result (str): Answer text.
+        - source (str): Question text.
+        - stage (int): Stage of the check (0, 1, or 2).
+
+        Returns:
+        dict: Result of the check containing
+    """
+
+    source_entity = extract_entities(source)
+    check_entity = extract_entities(result)
```

Comment:
Let's rework this `stage` variable. You are implementing a state machine here. So give the states meaningful names, e.g. TEXT_SIMILARITY_BASIC = 0, TEXT_SIMILARITY_BY_METRICS = 1, TEXT_SIMILARITY_BY_LLM = 2. Something like that. 

If function callers must know what these states mean then define them at the top of this file so they are obvious. 

Alternatively, is there a way to hide all of this from function callers? It seems like implementation detail to me, but I haven't been part of the design conversation.